include/device.h

```diff
@@ -498,6 +498,33 @@ device_required_handles_get(const struct device *dev,
  */
 __syscall const struct device *device_get_binding(const char *name);
 
+/**
+ * @brief Retrieve the device structure for a driver by name.
+ *
+ * This is the implementation underlying device_from_name(), without
+ * the overhead of a syscall wrapper.
+ *
+ * @param name device name to search for.  A null pointer, or a pointer to an
+ * empty string, will cause NULL to be returned.
+ *
+ * @return pointer to device structure; NULL if the name is not
+ * acceptable or no device is associated with the name.
+ */
+const struct device *z_device_from_name(const char *name);
+
+/**
+ * @brief Retrieve the device structure for a driver by name.
+ *
+ * This differs from device_get_binding() in that it does not verify
+ * that the device is ready to use.
```

Comment:
It would be nice to consider deprecating device_get_binding entirely. Benefits:

1. Out of tree users will find out about DEVICE_DT_GET and this new function, assuming we write the deprecation message well enough
2. device_get_binding() has always been a terrible name that causes new users confusion every single time in my experience. They think they've really "bound" to something in some stateful way instead of just done a lookup by name + a status check.

Do you have such plans?

Reply:
I agree, but no, I don't have plans to do that.